### PR TITLE
Fix SecurityException when HDFS Repository used against HA Namenodes

### PR DESCRIPTION
**Pull request by jbaiera**
When running the HDFS Repository Plugin against an HDFS cluster secured with Kerberos and configured with an HA Namenode topology, the repository will throw security exceptions and become unusable when the active Namenode fails over to the standby Namenode. The internal failover client code for HDFS attempts to establish a new connection to a different Namenode under the hood, but is unable to as it lacks the permissions to do so. After the client is created, all regular operations are done with restricted permissions to further police the client's behavior. These permissions do not allow the re-negotiation of the Namenode connection, which happens inside the client code.

This pull request attempts to sense when HA HDFS settings are present and remove permission restrictions during regular execution.

* Adding integration test for HA-Namenode-Enabled HDFS, regular and secured.
* Upgrade the MiniHDFS fixture to stand up an HA Namenode topology when configured with a nameservice.
* Add debug statements to log the configurations being added to the underlying HDFS configuration at repository creation time.
* Include a failing integration test for running an HDFS repository against an HDFS cluster, transitioning active Namenodes between repository operations.
* Include configurations for a new HA HDFS Fixture, and a Secure HA HDFS Fixture in the build script.
* Include new integration test tasks for both HA fixtures.
* HDFS Repository will still be subject to the permissions in the policy file but will not restrict them further when executing against HA configurations.
* Move doPrivileged blocks to HdfsSecurityContext so that it can decide when to restrict permissions.

### CODE REVIEW

**Comment by risdenk** [comment]
@jbaiera I got the following exception on my Mac.

```
...
    java.lang.RuntimeException: Unable to bind on specified streaming port in secure context. Needed 0, got 62869
        at org.apache.hadoop.hdfs.server.datanode.SecureDataNodeStarter.getSecureResources(SecureDataNodeStarter.java:112)
        at org.apache.hadoop.hdfs.MiniDFSCluster.startDataNodes(MiniDFSCluster.java:1485)
        at org.apache.hadoop.hdfs.MiniDFSCluster.initMiniDFSCluster(MiniDFSCluster.java:847)
        at org.apache.hadoop.hdfs.MiniDFSCluster.<init>(MiniDFSCluster.java:483)
        at org.apache.hadoop.hdfs.MiniDFSCluster$Builder.build(MiniDFSCluster.java:442)
        at hdfs.MiniHDFS.main(MiniHDFS.java:119)
```

I think it is related to https://issues.apache.org/jira/browse/HDFS-9213

I had previously run into this and worked around it by copying the `SecureDataNodeStarter` class into the project and applying the patch from HDFS-9213. An example of that class is here: https://github.com/risdenk/elasticsearch/commit/fa3892cd8aaadb7cd4cc985f456be21527864cc3

I have not tried that with your PR yet.

**Comment by jbaiera** [comment]
That's strange, it seems to have still passed CI. There may be something on your local environment that is keeping the minicluster from acquiring those privileged ports. It seems based on the issue you linked that this is a known issue with MiniDFSCluster despite the fact that we're already configuring it it not use the privileged ports...

I did some digging and found the following in the MiniDFSCluster class (excuse the decompiled variable names):

```
if(UserGroupInformation.isSecurityEnabled() && conf.get("dfs.data.transfer.protection") == null) {
    try {
        secureResources = SecureDataNodeStarter.getSecureResources(dnConf);
    } catch (Exception var27) {
        var27.printStackTrace();
    }
}
```

Setting `dfs.data.transfer.protection` would enable SASL authentication for the data xfer protocol and workaround the minicluster's issues with privileged ports. I'll update the tests to use SASL auth for the data transfer protocol.

**Comment by jbaiera** [comment]
Looks like CI has been broken on this PR. Gave it a kick.

**Comment by risdenk** [comment]
@jbaiera I looked through the test logs and don't see the integTestSecure* tests in the logs? Is Jenkins running the secure tests?

With the latest change I see:
```
2017-11-04 11:17:37,443 WARN  [main] server.KerberosAuthenticationHandler (KerberosAuthenticationHandler.java:init(284)) - HTTP principal: [hdfs/hdfs.build.elastic.co@BUILD.ELASTIC.CO] is invalid for SPNEGO!
```

The SASL auth requires HTTP/_HOST principals usually for SPNEGO. 

As far as the secure ports, I don't run the tests as a root or privileged user so the HDFS code can't bind to those low port numbers. The JIRA I pointed out I think is just bailing out early making it hard to test without a privileged user.

**Comment by jbaiera** [comment]
Seems the CI build is not running the secure tests most likely due to them lacking a vagrant installation. There are definitely other CI boxes that do have vagrant installed, they're just a different class. I'll double check to make sure that these tests are getting run on those.

Regarding the workaround - I'm hesitant to patch any of the HDFS tools for testing purposes. Even though it's a test-related artifact, I'd rather make sure that our tests work as-is to make future upgrades simpler.

>The SASL auth requires HTTP/_HOST principals usually for SPNEGO.

We're not really using any of the HTTP facilities for HDFS in the tests at the moment. If this is causing errors in your build other than just warning messages, I can take a look at configuring the HTTP principals, but if not I'd rather not delay the PR further if it can be avoided.

**Comment by jbaiera** [comment]
@rjernst Can you take a look at this when you're available?

**Comment by risdenk** [comment]
@jbaiera I haven't been able to get a secure test run to succeed. If there is a CI build that you have seen pass with the secure tests that would be great. I'm not sure the warning is related just wanted to point it out. I agree about not wanting to patch HDFS code.

Regardless, the changes look reasonable to me.

**Comment by jbaiera** [comment]
@risdenk are you still seeing issues with privileged ports or is it a different problem? Are there any stack traces you can post here?

**Comment by risdenk** [comment]
@jbaiera - I couldn't find a stack trace the last few times I tried. I would have provided it if I could. Everything looks good just comes back with a failure. Even with `--debug` there isn't a stacktrace that I could find. I'll try again in the next few days. It could be something on my machine if on other environments the tests pass. 

As a side note, what is the command you run to run the tests (without running the whole ES test suite)?

**Comment by risdenk** [comment]
@jbaiera - I might have figured out my issue. I should have an update later tonight. I might have been running from the wrong directory instead of at the top level of elasticsearch directory structure.

**Comment by risdenk** [comment]
@jbaiera - I haven't been able to get the tests to pass on my Mac (I run into a unable to start HDFS fixture). I ran on a Linux box and all the tests pass. I'll figure out the test issue some other time on my Mac. Sorry for the confusion.

**Comment by jbaiera** [comment]
> I run into a unable to start HDFS fixture

@risdenk You mentioned that you are not getting any exceptions. The build will abort the fixture if it takes longer than 30 seconds to standup and be initialized (write a port and pid file at the end of initialization). Are you seeing this timeout?

**Comment by risdenk** [comment]
@jbaiera definitely looks like 30 seconds in the logs. It doesn't say that it timeout out though. 

The below log is with `--info`. Even with debug the logs are more verbose but doesn't say anything about a timeout.
```
...
  [log]
    2017-11-07 21:22:25,500 WARN  [main] util.NativeCodeLoader (NativeCodeLoader.java:<clinit>(62)) - Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
...
    2017-11-07 21:22:54,918 INFO  [main] mortbay.log (Slf4jLog.java:info(67)) - Started HttpServer2$SelectChannelConnectorWithSafeStartup@localhost:63639
-----------------------------------------
:plugins:repository-hdfs:secureHdfsFixture FAILED
:plugins:repository-hdfs:secureHdfsFixture (Thread[main,5,main]) completed. Took 30.414 secs.
```

**Comment by jbaiera** [comment]
@risdenk re-running with `--stacktrace` might help as well. I think you might be running into an issue with MacOS and it's localhost domain name resolution. The last time people were reporting the fixture not working on Mac it was fixed by mapping `127.0.0.1` and `::1` to `localhost` in their /etc/hosts file.

**Comment by risdenk** [comment]
Both `127.0.0.1` and `::1` are mapped to `localhost` in my /etc/hots. It looks like the HDFS fixture test is starting correctly just times out. Is there an easy way to increase the timeout to see if that helps?

**Comment by risdenk** [comment]
For reference I modified https://github.com/elastic/elasticsearch/blob/master/buildSrc/src/main/groovy/org/elasticsearch/gradle/test/AntFixture.groovy#L132 to 60 seconds from 30. The test seem to get further. Currently looking into why the secure tests seem to be stalling some when talking to HDFS.

**Comment by risdenk** [comment]
@jbaiera - Finally got closer to figuring out what is happening:

* HDFS is taking longer than 30 seconds for me to startup. 
  * Work around: https://github.com/elastic/elasticsearch/blob/master/buildSrc/src/main/groovy/org/elasticsearch/gradle/test/AntFixture.groovy#L132 to 60 seconds from 30. 
* My network seems to have misconfigured IPV6 support
  * I've disabled IPV6 for my network adapter and have `::1 localhost` in my /etc/hosts.
  * At least one of the `RepositoryHdfsClientYamlTestSuiteIT` tests fail due to a 30 second timeout without the work around below.
  * Work around: `export _JAVA_OPTIONS='-Djava.net.preferIPv4Stack=true'` before running tests.

At this point I have the following tests working:
* `./gradlew integTestSecure --info`
* `./gradlew integTestSecureHa --info`

Thanks so much for your patience on this.

**Comment by jbaiera** [comment]
@risdenk Hey man, yeah, no problem! We also appreciate your feedback and work on verifying the fix on this as well.

File: test/fixtures/hdfs-fixture/src/main/java/hdfs/MiniHDFS.java [hunk 1]
```diff
@@ -91,6 +94,8 @@ public static void main(String[] args) throws Exception {
             cfg.set(DFSConfigKeys.DFS_NAMENODE_ACLS_ENABLED_KEY, "true");
             cfg.set(DFSConfigKeys.DFS_BLOCK_ACCESS_TOKEN_ENABLE_KEY, "true");
             cfg.set(DFSConfigKeys.IGNORE_SECURE_PORTS_FOR_TESTING_KEY, "true");
+            cfg.set(HdfsClientConfigKeys.DFS_DATA_TRANSFER_PROTECTION_KEY, "authentication");
+            cfg.set(DFSConfigKeys.DFS_ENCRYPT_DATA_TRANSFER_KEY, "false");
```
Comment (ywelsch) [inline test/fixtures/hdfs-fixture/src/main/java/hdfs/MiniHDFS.java]: Should we also test this with encryption enabled? Are all required jars for that on the classpath?

File: plugins/repository-hdfs/build.gradle [hunk 2]
```diff
@@ -165,6 +195,29 @@ if (fixtureSupported) {
   // just tests that the plugin loads
   integTestRunner.systemProperty 'tests.rest.suite', 'hdfs_repository/10_basic'
 }
+integTestRunner.exclude('**/Ha*TestSuiteIT.class') // Ignore HA integration tests.
```
Comment (ywelsch) [inline plugins/repository-hdfs/build.gradle]: mention that they are executed below as part of `integTestHa`

File: plugins/repository-hdfs/build.gradle [hunk 3]
```diff
@@ -138,6 +146,28 @@ task secureHdfsFixture(type: org.elasticsearch.gradle.test.AntFixture) {
   args miniHDFSArgs.toArray()
 }
 
+task secureHaHdfsFixture(type: org.elasticsearch.gradle.test.AntFixture) {
```
Comment (ywelsch) [inline plugins/repository-hdfs/build.gradle]: Can you do some code reuse here? For example

```
for (String fixture : ['secureHdfsFixture', 'secureHaHdfsFixture']) {
  tasks.create(fixture, org.elasticsearch.gradle.test.AntFixture) {
    ...
    if (fixture.equals('secureHaHdfsFixture')) {
      miniHDFSArgs += ["-Dha-nameservice=ha-hdfs"]
    }
    ...
  }
}
```

File: plugins/repository-hdfs/build.gradle [hunk 4]
```diff
@@ -205,10 +258,72 @@ if (secureFixtureSupported) {
...
+          " " + "-Djava.security.krb5.conf=${krb5conf}" +
+          " " + System.getProperty('tests.jvm.argline', '')
+  final String jvmArgs
+  if (project.rootProject.ext.javaVersion == JavaVersion.VERSION_1_9) {
+    jvmArgs = baseClusterJvmArgs + " " + '--add-opens=java.security.jgss/sun.security.krb5=ALL-UNNAMED'
+  } else {
+    jvmArgs = baseClusterJvmArgs
+  }
+
+  project.integTestSecureHa.clusterConfig.jvmArgs = jvmArgs
```
Comment (ywelsch) [inline plugins/repository-hdfs/build.gradle]: same comment about code reuse for everything in the afterEvaluate up to this point.

File: plugins/repository-hdfs/build.gradle [hunk 5]
```diff
@@ -205,10 +258,72 @@ if (secureFixtureSupported) {
...
+  project.integTestSecureHaRunner.onNonEmptyWorkDirectory 'ignore'
+}
+
+RestIntegTestTask integTestSecureHa = project.tasks.create('integTestSecureHa', RestIntegTestTask.class) {
+  description = "Runs rest tests against an elasticsearch cluster with HDFS configured with HA Namenode and secured by MIT Kerberos."
+}
+
+// set the keytab file somewhere we can access it from the test. In this case, place it into the temporary directory:
+project.integTestSecureHaRunner.doFirst {
+  Path tempDir = project.buildDir.toPath().resolve("testrun").resolve("integTestSecureHaRunner").resolve("J0").resolve("temp")
```
Comment (ywelsch) [inline plugins/repository-hdfs/build.gradle]: ah...this is so hacked.... how about adding the folder with the keytab file to the test classpath instead?

File: plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/HaHdfsFailoverTestSuiteIT.java [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.test.rest.ESRestTestCase;
+import org.junit.Assert;
+
+/**
+ * Integration test that runs against an HA-Enabled HDFS instance
+ */
+public class HaHdfsFailoverTestSuiteIT extends ESRestTestCase {
+
+    private static final Logger LOGGER = Loggers.getLogger(HaHdfsFailoverTestSuiteIT.class);
```
Comment (ywelsch) [inline plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/HaHdfsFailoverTestSuiteIT.java]: no need to instantiate own logger here, there is one inherited from ESTestCase.

File: plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/HaHdfsFailoverTestSuiteIT.java [hunk 7]
```diff
@@ -0,0 +1,289 @@
...
+                        "\"conf.dfs.ha.namenodes.ha-hdfs\": \"nn1,nn2\"," +
+                        "\"conf.dfs.namenode.rpc-address.ha-hdfs.nn1\": \"localhost:10001\"," +
+                        "\"conf.dfs.namenode.rpc-address.ha-hdfs.nn2\": \"localhost:10002\"," +
+                        "\"conf.dfs.client.failover.proxy.provider.ha-hdfs\": " +
+                            "\"org.apache.hadoop.hdfs.server.namenode.ha.ConfiguredFailoverProxyProvider\"" +
+                    "}" +
+                "}",
+                Charset.defaultCharset()), contentHeader);
+
+            Assert.assertEquals(200,response.getStatusLine().getStatusCode());
```
Comment (ywelsch) [inline plugins/repository-hdfs/src/test/java/org/elasticsearch/repositories/hdfs/HaHdfsFailoverTestSuiteIT.java]: missing space ;)

**Review by ywelsch — CHANGES_REQUESTED** [review]
Left minor comments. Overall change looks good!

**Comment by jbaiera** [comment]
Thanks @ywelsch, I collapsed a good chunk of the build file down. Let me know if we're all set to go on this.

File: test/fixtures/hdfs-fixture/src/main/java/hdfs/MiniHDFS.java [hunk 8]
```diff
@@ -91,6 +94,8 @@ public static void main(String[] args) throws Exception {
             cfg.set(DFSConfigKeys.DFS_NAMENODE_ACLS_ENABLED_KEY, "true");
             cfg.set(DFSConfigKeys.DFS_BLOCK_ACCESS_TOKEN_ENABLE_KEY, "true");
             cfg.set(DFSConfigKeys.IGNORE_SECURE_PORTS_FOR_TESTING_KEY, "true");
+            cfg.set(HdfsClientConfigKeys.DFS_DATA_TRANSFER_PROTECTION_KEY, "authentication");
```
Comment (ywelsch) [inline test/fixtures/hdfs-fixture/src/main/java/hdfs/MiniHDFS.java]: is this still needed if we have `dfs.encrypt.data.transfer` set to true? According to the docs, one should supersede the other, see https://hadoop.apache.org/docs/r2.6.0/hadoop-project-dist/hadoop-hdfs/hdfs-default.xml

File: plugins/repository-hdfs/src/test/resources/rest-api-spec/test/secure_hdfs_repository/30_snapshot_readonly.yml [hunk 9]
```diff
@@ -16,6 +16,7 @@
             security:
               principal: "elasticsearch@BUILD.ELASTIC.CO"
             readonly: true
+            conf.dfs.data.transfer.protection: "authentication"
```
Comment (ywelsch) [inline plugins/repository-hdfs/src/test/resources/rest-api-spec/test/secure_hdfs_repository/30_snapshot_readonly.yml]: is this setting still needed with encryption being enabled? (see also all other occurrences of this)

**Review by ywelsch — APPROVED** [review]
Thanks, LGTM. I've left one small suggestion, which you can address at will. Please also add version labels to the PR.

**Comment by jbaiera** [comment]
Thanks all, I'll work on backporting this to the appropriate branches. Will update the tags to reflect which ones receive the fix.

**Comment by risdenk** [comment]
Thanks @jbaiera!

**Comment by risdenk** [comment]
@jbaiera - I just saw ES 6.1 was released. It looks like the v6.1.x tag was missed for this? 

```
git branch -r --contains f1ba986
  origin/6.1
```

I also don't see a v6.1.0 tag on github to check if the tag has the commit.

**Comment by risdenk** [comment]
I did some quick sleuthing and found that PR 27652 is in the release notes for 6.1.0 (https://www.elastic.co/guide/en/elasticsearch/reference/current/release-notes-6.1.0.html) and is the next commit (ea588e551a2fac827b71d4633ef28b07c3167aae) on the 6.1 branch after commit f1ba986 on this issue.

I think that means that this is in ES 6.1.0 and isn't in the release notes.

**Comment by jbaiera** [comment]
@risdenk Thanks for pointing that out. It looks like this slipped into 6.1.0 instead of in 6.1.1. I'll see about getting an update to the release notes out.

**Pull request merged**